caikit/runtime/client/remote_module_base.py

```diff
@@ -0,0 +1,496 @@
+        request_protobuf_class = request_dm_class.get_proto_class()
+
+        # Get the response types
+        response_dm_class = DataBase.get_class_for_name(method.response_dm_name)
+        response_protobuf_class = response_dm_class.get_proto_class()
+
+        # Get the RPC route
+        grpc_route = get_grpc_route_name(service_type, method.rpc_name)
+
+        # Construct the service_rpc and serializers
+        if method.input_streaming and method.output_streaming:
+            service_rpc = self._grpc_channel.stream_stream(
+                grpc_route,
+                request_serializer=request_protobuf_class.SerializeToString,
+                response_deserializer=response_protobuf_class.FromString,
+            )
+        elif method.input_streaming:
+            service_rpc = self._grpc_channel.stream_unary(
+                grpc_route,
+                request_serializer=request_protobuf_class.SerializeToString,
+                response_deserializer=response_protobuf_class.FromString,
+            )
+        elif method.output_streaming:
+            service_rpc = self._grpc_channel.unary_stream(
+                grpc_route,
+                request_serializer=request_protobuf_class.SerializeToString,
+                response_deserializer=response_protobuf_class.FromString,
+            )
+        else:
+            service_rpc = self._grpc_channel.unary_unary(
+                grpc_route,
+                request_serializer=request_protobuf_class.SerializeToString,
+                response_deserializer=response_protobuf_class.FromString,
+            )
+
+        # Construct request object
+        if method.input_streaming:
+            # Bind the args and kwargs to the signature for parsing. Use None for the self argument
+            bound_args = method.signature.method_signature.bind(None, *args, **kwargs)
+            bound_args.arguments.pop("self")
+
+            # Gather all iterable parameters as these should be streamed
+            streaming_kwargs = OrderedDict()
+            for name in self._get_streaming_arguments(**bound_args.arguments):
+                streaming_kwargs[name] = bound_args.arguments.pop(name)
+
+            def input_stream_parser():
+                """Helper function to iterate over a datastream and stream requests"""
+                for stream_tuple in DataStream.zip(*streaming_kwargs.values()):
+                    stream_arguments = copy.deepcopy(bound_args)
+                    for streaming_key, sub_value in zip(
+                        streaming_kwargs.keys(), stream_tuple
+                    ):
+                        stream_arguments.arguments[streaming_key] = sub_value
+
+                    yield request_dm_class(
+                        *stream_arguments.args, **stream_arguments.kwargs
+                    ).to_proto()
+
+            grpc_request = input_stream_parser()
+        else:
+            # If not streaming then construct a simple request
```

Comment:
So much simpler. Love the handling for streaming out-of-the-box though!

Reply:
Yeah, the nonstreaming/nontraining implementation was very neat..... I think adding support for streaming was a must/hard requirement but it does muddle things